File: v1/src/main/java/com/api/v1/customer/domain/Customer.java
package com.api.v1.customer.domain;

import com.api.v1.person.domain.Person;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.time.ZoneId;
import java.util.UUID;

@Getter
@NoArgsConstructor
@Document(collection = "v1_customers")
public class Customer {

    @Id
    private UUID id;
    private Person person;
    private Instant createdAt;
    private ZoneId createdAtZone;
    private Instant modifiedAt;
    private ZoneId modifiedAtZone;

    private Customer(Person person) {
        this.id = UUID.randomUUID();
        this.person = person;
        this.createdAt = Instant.now();
        this.createdAtZone = ZoneId.systemDefault();
    }

    public static Customer of(Person person) {
        return new Customer(person);
    }

    public void modify(Person person) {
        this.person = person;
        modifiedAt = Instant.now();
        modifiedAtZone = ZoneId.systemDefault();
    }

}

File: kotlin/v3/src/main/kotlin/com/api/v3/customers/domain/Customer.kt
package com.api.v3.customers.domain

import com.api.v3.persons.domain.Person
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.UUID

@Document
class Customer {

    @Id
    var id: UUID = UUID.randomUUID()
    var person: Person
    var createdAt: LocalDateTime = LocalDateTime.now()
    var createdAtZone: ZoneId = ZoneId.systemDefault()

    private constructor(person: Person) {
        this.person = person
    }

    companion object {
        fun of(person: Person): Customer {
            return Customer(person)
        }
    }

}